pkg/kademlia/kademlia_test.go

```diff
@@ -0,0 +1,395 @@
+		addr := test.RandomAddressAt(base, i)
+		addOne(t, kad, ab, addr)
+	}
+	waitConns(t, &conns, 0)
+
+	// add one peer in a bin higher (unsaturated) and expect one connection
+	addr := test.RandomAddressAt(base, 6)
+	addOne(t, kad, ab, addr)
+
+	waitConns(t, &conns, 1)
+	atomic.StoreInt32(&conns, 0)
+
+	// again, one bin higher
+	addr = test.RandomAddressAt(base, 7)
+	addOne(t, kad, ab, addr)
+
+	waitConns(t, &conns, 1)
+	atomic.StoreInt32(&conns, 0)
+
+	// this is in order to hit the `if size < 2` in the saturation func
+	removeOne(kad, peers[2])
+	waitConns(t, &conns, 1)
+}
+
+func TestBackoff(t *testing.T) {
+	// cheat and decrease the timer
+	defer func(t time.Duration) {
+		*kademlia.TimeToRetry = t
+	}(*kademlia.TimeToRetry)
+
+	*kademlia.TimeToRetry = 500 * time.Millisecond
+
+	var (
+		conns int32 // how many connect calls were made to the p2p mock
+
+		p2p = p2pmock.New(p2pmock.WithConnectFunc(func(_ context.Context, addr ma.Multiaddr) (swarm.Address, error) {
+			_ = atomic.AddInt32(&conns, 1)
+			return swarm.ZeroAddress, nil
+		}))
+		base, kad, ab = newTestKademlia(p2p, nil)
+	)
+
+	// add one peer, wait for connection
+	addr := test.RandomAddressAt(base, 1)
+	addOne(t, kad, ab, addr)
+
+	waitConns(t, &conns, 1)
+	atomic.StoreInt32(&conns, 0)
+
+	// remove that peer
+	removeOne(kad, addr)
+
+	// wait for 100ms, add another peer, expect just one more connection
+	time.Sleep(100 * time.Millisecond)
+	addr = test.RandomAddressAt(base, 1)
+	addOne(t, kad, ab, addr)
+
+	waitConns(t, &conns, 1)
+	atomic.StoreInt32(&conns, 0)
+
+	// wait for another 400ms, add another, expect 2 connections
+	time.Sleep(400 * time.Millisecond)
+	addr = test.RandomAddressAt(base, 1)
+	addOne(t, kad, ab, addr)
+
+	waitConns(t, &conns, 2)
+}
+
+func TestMarshal(t *testing.T) {
+	var (
+		p2p = p2pmock.New(p2pmock.WithConnectFunc(func(_ context.Context, addr ma.Multiaddr) (swarm.Address, error) {
+			return swarm.ZeroAddress, nil
+		}))
+		_, kad, ab = newTestKademlia(p2p, nil)
+	)
+	a := test.RandomAddress()
+	addOne(t, kad, ab, a)
+	_, err := kad.MarshalJSON()
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func newTestKademlia(p2p p2p.Service, f func(bin, depth uint8, peers *pslice.PSlice) bool) (swarm.Address, *kademlia.Kad, addressbook.Interface) {
+	var (
+		base   = test.RandomAddress()                                                                                     // base address
+		logger = logging.New(ioutil.Discard, 0)                                                                           // logger
+		ab     = addressbook.New(mockstate.NewStateStore())                                                               // address book
+		kad    = kademlia.New(kademlia.Options{Base: base, AddressBook: ab, P2P: p2p, Logger: logger, SaturationFunc: f}) // kademlia instance
+	)
+	return base, kad, ab
+}
+
+func removeOne(k *kademlia.Kad, peer swarm.Address) {
+	k.Disconnected(peer)
+}
+
+const underlayBase = "/ip4/127.0.0.1/tcp/7070/dns/"
```

Comment:
I am curious, why dns protocol?

Reply:
had to brute force something that will decode correctly when creating a new multiaddr when constructing it below

Reply:
We are using p2p protocol in addresses, dns may be confusing for someone that reads it.

Reply:
Is it ok to change dns to p2p?

Reply:
tried. doesn't work. if we want to have logging in the test that would be possible to debug (i.e. to see which address failed), then we have to hack it this way or we'd have to spend more time to hash the swarm addresses into whatever base-something that the p2p layer uses for it to be compatible with the `p2p` addressing scheme, since the multiaddr constructor verifies this when you call it with the ma string.

Reply:
That is very unfortunate. OK, let's leave it as it is now.